From 59c9333998928feadef842d75c65625097a8a0e1 Mon Sep 17 00:00:00 2001
From: Sam Ravnborg <srn@skov.dk>
Date: Wed, 15 Feb 2017 23:27:09 +0100
Subject: [PATCH] drop arm9-cpu support for dataflash + nandflash

The current HW do not have dataflash / nandflash.
Drop all code touching this, to keep changes lower

The patch also enable more debugging
---
 arch/arm/lib/board.c           |   6 --
 board/skov/arm9-cpu/arm9-cpu.c |  83 +++------------------------
 common/Makefile                |   3 -
 common/cmd_nand.c              |  10 ----
 common/env_dataflash.c         |  31 ----------
 common/env_flash.c             | 127 ++---------------------------------------
 common/env_skov_compat.c       |  89 -----------------------------
 common/main.c                  |  22 -------
 drivers/mtd/cfi_flash.c        |   9 +--
 drivers/mtd/dataflash.c        |  29 ----------
 include/configs/arm9-cpu.h     |  40 +++++++------
 include/environment.h          |  25 +-------
 tools/envcrc.c                 |   9 ---
 13 files changed, 38 insertions(+), 445 deletions(-)
 delete mode 100644 common/env_skov_compat.c

diff --git a/arch/arm/lib/board.c b/arch/arm/lib/board.c
index 86f681c..83222fa 100644
--- a/arch/arm/lib/board.c
+++ b/arch/arm/lib/board.c
@@ -71,9 +71,6 @@ ulong monitor_flash_len;
 #ifdef CONFIG_HAS_DATAFLASH
 extern int  AT91F_DataflashInit(void);
 extern void dataflash_print_info(void);
-#ifdef CONFIG_ARM9_CPU 
-extern void check_240(void);
-#endif
 #endif
 
 #if defined(CONFIG_HARD_I2C) || \
@@ -553,9 +550,6 @@ void board_init_r(gd_t *id, ulong dest_addr)
 #ifdef CONFIG_HAS_DATAFLASH
 	AT91F_DataflashInit();
 	dataflash_print_info();
-#ifdef CONFIG_ARM9_CPU
-	check_240(); /**< Martin Bjærre Is the board running 240Mhz? */
-#endif
 #endif
 
 	/* initialize environment */
diff --git a/board/skov/arm9-cpu/arm9-cpu.c b/board/skov/arm9-cpu/arm9-cpu.c
index 2a1dc15..4b6409c 100644
--- a/board/skov/arm9-cpu/arm9-cpu.c
+++ b/board/skov/arm9-cpu/arm9-cpu.c
@@ -115,57 +115,6 @@ int board_mmc_getcd(struct mmc *mmc)
 }
 #endif
 
-#ifdef CONFIG_CMD_NAND
-static void at91sam9263ek_nand_hw_init(void)
-{
-	unsigned long csa;
-	at91_smc_t    *smc    = (at91_smc_t *) ATMEL_BASE_SMC0;
-	at91_matrix_t *matrix = (at91_matrix_t *) ATMEL_BASE_MATRIX;
-	at91_pmc_t    *pmc    = (at91_pmc_t *) ATMEL_BASE_PMC;
-
-	/* Enable CS3 */
-	csa = readl(&matrix->csa[0]) | AT91_MATRIX_CSA_EBI_CS3A;
-	writel(csa, &matrix->csa[0]);
-
-	/* Enable CS3 */
-
-	/* Configure SMC CS3 for NAND/SmartMedia */
-	writel(AT91_SMC_SETUP_NWE(1) | AT91_SMC_SETUP_NCS_WR(0) |
-		AT91_SMC_SETUP_NRD(1) | AT91_SMC_SETUP_NCS_RD(0),
-		&smc->cs[3].setup);
-
-	writel(AT91_SMC_PULSE_NWE(3) | AT91_SMC_PULSE_NCS_WR(3) |
-		AT91_SMC_PULSE_NRD(3) | AT91_SMC_PULSE_NCS_RD(3),
-		&smc->cs[3].pulse);
-
-	writel(AT91_SMC_CYCLE_NWE(5) | AT91_SMC_CYCLE_NRD(5),
-		&smc->cs[3].cycle);
-	writel(AT91_SMC_MODE_RM_NRD | AT91_SMC_MODE_WM_NWE |
-		AT91_SMC_MODE_EXNW_DISABLE |
-#ifdef CONFIG_SYS_NAND_DBW_16
-		       AT91_SMC_MODE_DBW_16 |
-#else /* CONFIG_SYS_NAND_DBW_8 */
-		       AT91_SMC_MODE_DBW_8 |
-#endif
-		       AT91_SMC_MODE_TDF_CYCLE(2),
-		&smc->cs[3].mode);
-
-	writel(1 << ATMEL_ID_PIOA | 1 << ATMEL_ID_PIOCDE,
-		&pmc->pcer);
-	
-	/* Rose/HBS: We have R/B on pioB */
-	writel(1 << ATMEL_ID_PIOA | 1 << ATMEL_ID_PIOB | 1 << ATMEL_ID_PIOCDE,
-			&pmc->pcer);		
-
-	/* Configure RDY/BSY */
-	at91_set_gpio_input(CONFIG_SYS_NAND_READY_PIN, 1);
-
-	/* Enable NandFlash */
-	//at91_set_gpio_output(CONFIG_SYS_NAND_ENABLE_PIN, 1);
-	at91_set_gpio_output(CONFIG_SYS_NAND_ENABLE_PIN, 0); /**< Rose/HBS: stygt hack */
-}
-#endif
-
 #ifdef CONFIG_MACB
 static void at91sam9263ek_macb_hw_init(void)
 {
@@ -402,11 +351,13 @@ static void at91sam9263_lcd_hw_init(void)
 		 * This is the old interface board, as the pin is NC (using Seiko display)
 		 * Do nothing
 		 */
+		printf("Display: Seiko\n");
 	}
 	else {
 		/**
 		 * This is the new interface board, as the pin is pulled down (using Logic display)
 		 */
+		printf("Display: LTTD800480070\n");
 		panel_info = panel_info_logic;
 		/* enable contrast PWM(1) */
 		writel(1<<AT91_ID_PWM, &pmc->pcer); 
@@ -439,20 +390,6 @@ static void at91sam9263_lcd_hw_init(void)
 #endif
 }
 
-#if defined(CONFIG_ARM9_CPU)
-void check_240(void)
-{
-	if (do_we_have_dataflash())	{
-#ifdef CONFIG_LTTD800480070	
-		if (!at91_get_gpio_value(AT91_PIN_PD2)) {
-			panel_info = panel_info_logic_240;
-		}
-#endif
-	}
-}
-#endif
-
-
 #ifdef CONFIG_LCD_INFO
 #include <nand.h>
 #include <version.h>
@@ -480,10 +417,6 @@ void lcd_show_board_info(void)
 	for (i = 0; i < CONFIG_NR_DRAM_BANKS; i++)
 		dram_size += gd->bd->bi_dram[i].size;
 	nand_size = 0;
-#ifdef CONFIG_CMD_NAND
-	for (i = 0; i < CONFIG_SYS_MAX_NAND_DEVICE; i++)
-		nand_size += nand_info[i].size;
-#endif
 #ifndef CONFIG_SYS_NO_FLASH
 	flash_size = 0;
 	for (i = 0; i < CONFIG_SYS_MAX_FLASH_BANKS; i++)
@@ -522,13 +455,6 @@ int board_init(void)
 	/* adress of boot parameters */
 	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
 
-#ifdef CONFIG_CMD_NAND
-	at91sam9263ek_nand_hw_init();
-#endif
-#ifdef CONFIG_HAS_DATAFLASH
-	at91_set_pio_output(AT91_PIO_PORTE, 20, 1);	/* select spi0 clock */
-	at91_spi0_hw_init(1 << 0);
-#endif
 #ifdef CONFIG_MACB
 	at91sam9263ek_macb_hw_init();
 #endif
@@ -585,6 +511,11 @@ int board_init(void)
 	return 0;
 }
 
+void enable_caches(void)
+{
+	printf("enable_caches\n");
+}
+
 int dram_init(void)
 {
 	/* ROSE: We will configure the memory size accourding the two values written by the bootstrap */ 
diff --git a/common/Makefile b/common/Makefile
index 61fe837..973f05a 100644
--- a/common/Makefile
+++ b/common/Makefile
@@ -61,9 +61,6 @@ COBJS-$(CONFIG_ENV_IS_IN_SPI_FLASH) += env_sf.o
 COBJS-$(CONFIG_ENV_IS_IN_REMOTE) += env_remote.o
 COBJS-$(CONFIG_ENV_IS_NOWHERE) += env_nowhere.o
 
-# Skov environment
-COBJS-$(CONFIG_ARM9_CPU) += env_skov_compat.o
-
 # command
 COBJS-$(CONFIG_CMD_AMBAPP) += cmd_ambapp.o
 COBJS-$(CONFIG_SOURCE) += cmd_source.o
diff --git a/common/cmd_nand.c b/common/cmd_nand.c
index c0826a3..9a5fe6f 100644
--- a/common/cmd_nand.c
+++ b/common/cmd_nand.c
@@ -584,11 +584,7 @@ int do_nand(cmd_tbl_t * cmdtp, int flag, int argc, char * const argv[])
 					 * This functionality is removed as it endangers bad block information
 					 *
 					 */
-#ifdef CONFIG_ARM9_CPU
-					printf("\n nand scrub functionality has been removed to protect bad block information on skov cpu__module\n");
-#else
 					opts.scrub = 1;
-#endif
 				else {
 					puts("scrub aborted\n");
 					return -1;
@@ -702,13 +698,7 @@ int do_nand(cmd_tbl_t * cmdtp, int flag, int argc, char * const argv[])
                                  * This functionality is removed as it endangers bad block information
                                  *
                                  */
-#ifdef CONFIG_ARM9_CPU
-				printf("\n nand write.oob functionality has been removed to protect bad block information.");
-				ret = -1;
-				size = 0;
-#else
 				ret = nand->write_oob(nand, off, &ops);
-#endif
 		} else if (raw) {
 			ret = raw_access(nand, addr, off, pagecount, read);
 		} else {
diff --git a/common/env_dataflash.c b/common/env_dataflash.c
index 3a7105c..d641f09 100644
--- a/common/env_dataflash.c
+++ b/common/env_dataflash.c
@@ -27,23 +27,11 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
-#ifdef CONFIG_ARM9_CPU
-extern env_t *env_ptr;
-#else
 env_t *env_ptr;
-#endif
 
-#ifdef CONFIG_ARM9_CPU
-char * env_name_spec_dataflash = "dataflash";
-#else
 char *env_name_spec = "dataflash";
-#endif
 
-#ifdef CONFIG_ARM9_CPU
-uchar env_get_char_spec_dataflash (int index)
-#else
 uchar env_get_char_spec(int index)
-#endif
 {
 	uchar c;
 
@@ -51,11 +39,7 @@ uchar env_get_char_spec(int index)
 			1, (char *)&c);
 	return c;
 }
-#ifdef CONFIG_ARM9_CPU
-void env_relocate_spec_dataflash (void)
-#else
 void env_relocate_spec(void)
-#endif
 {
 	char buf[CONFIG_ENV_SIZE];
 
@@ -68,11 +52,7 @@ void env_relocate_spec(void)
 #error No support for redundant environment on dataflash yet!
 #endif
 
-#ifdef CONFIG_ARM9_CPU
-int saveenv_dataflash(void)
-#else
 int saveenv(void)
-#endif
 {
 	env_t	env_new;
 	ssize_t	len;
@@ -97,11 +77,7 @@ int saveenv(void)
  * We are still running from ROM, so data use is limited.
  * Use a (moderately small) buffer on the stack
  */
-#ifdef CONFIG_ARM9_CPU
-int env_init_dataflash(void)
-#else
 int env_init(void)
-#endif
 {
 	ulong crc, len = ENV_SIZE, new = 0;
 	unsigned off;
@@ -109,15 +85,8 @@ int env_init(void)
 
 	if (gd->env_valid)
 		return 0;
-/**
- * @autor Søren Andersen
- * The Bootstrap had already initialized the flash
- */
-#if !defined(CONFIG_ARM9_CPU)		
 	AT91F_DataflashInit();	/* prepare for DATAFLASH read/write */
-#endif
 	/* read old CRC */
-
 	read_dataflash(CONFIG_ENV_ADDR + offsetof(env_t, crc),
 		sizeof(ulong), (char *)&crc);
 
diff --git a/common/env_flash.c b/common/env_flash.c
index f62225f..08a6ae6 100644
--- a/common/env_flash.c
+++ b/common/env_flash.c
@@ -47,45 +47,22 @@ DECLARE_GLOBAL_DATA_PTR;
 #error CONFIG_ENV_SIZE_REDUND should not be less then CONFIG_ENV_SIZE
 #endif /* CONFIG_ENV_SIZE_REDUND */
 
-#ifdef CONFIG_ARM9_CPU
-char *env_name_spec_flash = "Flash";
-#else
 char *env_name_spec = "Flash";
-#endif  /* CONFIG_ARM9_CPU */
 
 #ifdef ENV_IS_EMBEDDED
 env_t *env_ptr = &environment;
 
-#ifdef CONFIG_ARM9_CPU
-static env_t *flash_addr = (env_t *)CONFIG_ENV_ADDR_FLASH; /**< Martin Bjærre: Changed CONFIG_ENV_ADDR to CONFIG_ENV_ADDR_FLASH */
-#else
 static env_t *flash_addr = (env_t *)CONFIG_ENV_ADDR;
-#endif /* CONFIG_ARM9_CPU */
-
 #else /* ! ENV_IS_EMBEDDED */
 
-#ifdef CONFIG_ARM9_CPU
-extern env_t *env_ptr; /**< Martin Bjærre: env_t is now located in env_skov_compat. */
-#else
 env_t *env_ptr = (env_t *)CONFIG_ENV_ADDR;
-#endif /* CONFIG_ARM9_CPU */
 
-#ifdef CONFIG_ARM9_CPU
-static env_t *flash_addr = (env_t *)CONFIG_ENV_ADDR_FLASH;
-#else
 static env_t *flash_addr = (env_t *)CONFIG_ENV_ADDR;
-#endif /* CONFIG_ARM9_CPU */
-
 #endif /* ENV_IS_EMBEDDED */
 
 #if defined(CMD_SAVEENV) || defined(CONFIG_ENV_ADDR_REDUND)
-#ifdef CONFIG_ARM9_CPU
-//Rose/Mab: Changed CONFIG_ENV_ADDR to CONFIG_ENV_ADDR_FLASH
-static ulong end_addr = CONFIG_ENV_ADDR_FLASH + CONFIG_ENV_SECT_SIZE_FLASH - 1;
-#else
 /* CONFIG_ENV_ADDR is supposed to be on sector boundary */
 static ulong end_addr = CONFIG_ENV_ADDR + CONFIG_ENV_SECT_SIZE - 1;
-#endif /* CONFIG_ARM9_CPU */
 #endif /* CMD_SAVEENV || CONFIG_ENV_ADDR_REDUND */
 
 #ifdef CONFIG_ENV_ADDR_REDUND
@@ -97,11 +74,7 @@ static ulong end_addr_new = CONFIG_ENV_ADDR_REDUND + CONFIG_ENV_SECT_SIZE - 1;
 
 
 #ifdef CONFIG_ENV_ADDR_REDUND
-#ifdef CONFIG_ARM9_CPU
-int env_init_flash(void)
-#else
 int env_init(void)
-#endif /* CONFIG_ARM9_CPU */
 {
 	int crc1_ok = 0, crc2_ok = 0;
 
@@ -112,14 +85,9 @@ int env_init(void)
 	ulong addr1 = (ulong)&(flash_addr->data);
 	ulong addr2 = (ulong)&(flash_addr_new->data);
 
-#ifdef CONFIG_ARM9_CPU
-	crc1_ok = (crc32(0, flash_addr->data, ENV_SIZE_FLASH) == flash_addr->crc);
-	crc2_ok = (crc32(0, flash_addr_new->data, ENV_SIZE_FLASH) == flash_addr_new->crc);
-#else
 	crc1_ok = crc32(0, flash_addr->data, ENV_SIZE) == flash_addr->crc;
 	crc2_ok =
 		crc32(0, flash_addr_new->data, ENV_SIZE) == flash_addr_new->crc;
-#endif /* CONFIG_ARM9_CPU */
 	if (crc1_ok && !crc2_ok) {
 		gd->env_addr	= addr1;
 		gd->env_valid	= 1;
@@ -150,18 +118,14 @@ int env_init(void)
 }
 
 #ifdef CMD_SAVEENV
-#ifdef CONFIG_ARM9_CPU
-int saveenv_flash(void)
-#else
 int saveenv(void)
-#endif /* CONFIG_ARM9_CPU */
 {
 	env_t	env_new;
 	ssize_t	len;
 	char	*res, *saved_data = NULL;
 	char	flag = OBSOLETE_FLAG, new_flag = ACTIVE_FLAG;
 	int	rc = 1;
-#if defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) || !defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
+#if (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
         ulong   up_data = 0;
 #endif /* (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) */
 	debug("Protect off %08lX ... %08lX\n", (ulong)flash_addr, end_addr);
@@ -176,28 +140,16 @@ int saveenv(void)
 		goto done;
 
 	res = (char *)&env_new.data;
-#ifdef CONFIG_ARM9_CPU
-        len = hexport_r(&env_htab, '\0', &res, ENV_SIZE_FLASH, 0, NULL);
-#else
 	len = hexport_r(&env_htab, '\0', &res, ENV_SIZE, 0, NULL);
-#endif /* CONFIG_ARM9_CPU */
 	if (len < 0) {
 		error("Cannot export environment: errno = %d\n", errno);
 		goto done;
 	}
-#ifdef CONFIG_ARM9_CPU
-	env_new.crc	= crc32(0, env_new.data, ENV_SIZE_FLASH);
-#else
         env_new.crc     = crc32(0, env_new.data, ENV_SIZE);
-#endif /* CONFIG_ARM9_CPU */
 	env_new.flags	= new_flag;
 
-#if defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) || !defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
-#ifdef CONFIG_ARM9_CPU
-        up_data = end_addr_new + 1 - ((long)flash_addr_new + CONFIG_ENV_SIZE_FLASH);
-#else
+#if (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
 	up_data = end_addr_new + 1 - ((long)flash_addr_new + CONFIG_ENV_SIZE);
-#endif /* CONFIG_ARM9_CPU */
 	debug("Data to save 0x%lX\n", up_data);
 	if (up_data) {
 		saved_data = malloc(up_data);
@@ -207,20 +159,11 @@ int saveenv(void)
 			goto done;
 		}
 		memcpy(saved_data,
-#ifdef CONFIG_ARM9_CPU
-			(void *)((long)flash_addr_new + CONFIG_ENV_SIZE_FLASH), up_data);
-#else
 			(void *)((long)flash_addr_new + CONFIG_ENV_SIZE),
 			up_data);
-#endif /* CONFIG_ARM9_CPU */			
 		debug("Data (start 0x%lX, len 0x%lX) saved at 0x%p\n",
-#ifdef CONFIG_ARM9_CPU
-                        (long)flash_addr_new + CONFIG_ENV_SIZE_FLASH,
-                                                up_data, saved_data);
-#else
 			(long)flash_addr_new + CONFIG_ENV_SIZE,
 			up_data, saved_data);
-#endif /* CONFIG_ARM9_CPU */
 	}
 #endif /* (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) */
 	puts("Erasing Flash...");
@@ -243,21 +186,13 @@ int saveenv(void)
 	if (rc)
 		goto perror;
 
-#if defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) || !defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
+#if (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
 	if (up_data) { /* restore the rest of sector */
 		debug("Restoring the rest of data to 0x%lX len 0x%lX\n",
-#ifdef CONFIG_ARM9_CPU
-                        (long)flash_addr_new + CONFIG_ENV_SIZE_FLASH, up_data);
-#else
 			(long)flash_addr_new + CONFIG_ENV_SIZE, up_data);
-#endif /* CONFIG_ARM9_CPU */
 		if (flash_write(saved_data,
-#ifdef CONFIG_ARM9_CPU
-                                (long)flash_addr_new + CONFIG_ENV_SIZE_FLASH, up_data))
-#else				
                                 (long)flash_addr_new + CONFIG_ENV_SIZE,
                                 up_data))
-#endif /* CONFIG_ARM9_CPU */
 			goto perror;
 	}
 #endif /* (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) */
@@ -291,11 +226,7 @@ done:
 
 #else /* ! CONFIG_ENV_ADDR_REDUND */
 
-#ifdef CONFIG_ARM9_CPU
-int env_init_flash(void)
-#else
 int env_init(void)
-#endif /* CONFIG_ARM9_CPU */
 {
 	if (crc32(0, env_ptr->data, ENV_SIZE) == env_ptr->crc) {
 		gd->env_addr	= (ulong)&(env_ptr->data);
@@ -309,23 +240,15 @@ int env_init(void)
 }
 
 #ifdef CMD_SAVEENV
-#ifdef CONFIG_ARM9_CPU
-int saveenv_flash(void)
-#else
 int saveenv(void)
-#endif /* CONFIG_ARM9_CPU */
 {
 	env_t	env_new;
 	ssize_t	len;
 	int	rc = 1;
 	char	*res, *saved_data = NULL;
-#if defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) || !defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
+#if (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
 	ulong	up_data = 0;
-#ifdef CONFIG_ARM9_CPU
-        up_data = end_addr + 1 - ((long)flash_addr + CONFIG_ENV_SIZE_FLASH);
-#else
 	up_data = end_addr + 1 - ((long)flash_addr + CONFIG_ENV_SIZE);
-#endif /* CONFIG_ARM9_CPU */
 	debug("Data to save 0x%lx\n", up_data);
 	if (up_data) {
 		saved_data = malloc(up_data);
@@ -335,17 +258,9 @@ int saveenv(void)
 			goto done;
 		}
 		memcpy(saved_data,
-#ifdef CONFIG_ARM9_CPU
-                        (void *)((long)flash_addr + CONFIG_ENV_SIZE_FLASH), up_data); 
-#else
 			(void *)((long)flash_addr + CONFIG_ENV_SIZE), up_data);
-#endif /* CONFIG_ARM9_CPU */
 		debug("Data (start 0x%lx, len 0x%lx) saved at 0x%lx\n",
-#ifdef CONFIG_ARM9_CPU
-                        (ulong)flash_addr + CONFIG_ENV_SIZE_FLASH, up_data, (ulong)saved_data);
-#else
 			(ulong)flash_addr + CONFIG_ENV_SIZE, up_data, (ulong)saved_data);
-#endif /* CONFIG_ARM9_CPU */
 	}
 #endif	/* (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) */
 	debug("Protect off %08lX ... %08lX\n", (ulong)flash_addr, end_addr);
@@ -354,20 +269,12 @@ int saveenv(void)
 		goto done;
 	}	
 	res = (char *)&env_new.data;
-#ifdef CONFIG_ARM9_CPU
-        len = hexport_r(&env_htab, '\0', &res, ENV_SIZE_FLASH, 0, NULL);
-#else
 	len = hexport_r(&env_htab, '\0', &res, ENV_SIZE, 0, NULL);
-#endif /* CONFIG_ARM9_CPU */
 	if (len < 0) {
 		error("Cannot export environment: errno = %d\n", errno);
 		goto done;
 	}
-#ifdef CONFIG_ARM9_CPU
-        env_new.crc = crc32(0, env_new.data, ENV_SIZE_FLASH);
-#else
 	env_new.crc = crc32(0, env_new.data, ENV_SIZE);
-#endif /* CONFIG_ARM9_CPU */
 	puts("Erasing Flash...");
 
 	if (flash_sect_erase((long)flash_addr, end_addr))
@@ -375,34 +282,18 @@ int saveenv(void)
 
 	puts("Writing to Flash... ");
 
-#ifdef CONFIG_ARM9_CPU
-        rc = flash_write((char *)&env_new, (long)flash_addr, CONFIG_ENV_SIZE_FLASH);
-#else
 	rc = flash_write((char *)&env_new, (long)flash_addr, CONFIG_ENV_SIZE);
-#endif /* CONFIG_ARM9_CPU */
 
 	if (rc != 0)
 		goto perror;
-#if defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) || !defined(CONFIG_ARM9_CPU) && (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
+#if (CONFIG_ENV_SECT_SIZE > CONFIG_ENV_SIZE_FLASH)
 	if (up_data) {	/* restore the rest of sector */
 		debug("Restoring the rest of data to 0x%lx len 0x%lx\n",
-#ifdef CONFIG_ARM9_CPU
-                        (ulong)flash_addr + CONFIG_ENV_SIZE_FLASH, up_data);
-#else
 			(ulong)flash_addr + CONFIG_ENV_SIZE, up_data);
-#endif /* CONFIG_ARM9_CPU */
 
-
-#ifdef CONFIG_ARM9_CPU
-                if (flash_write(saved_data, (long)flash_addr + CONFIG_ENV_SIZE_FLASH, up_data)) {
-                        goto perror;
-                }
-#else
 		if (flash_write(saved_data, (long)flash_addr + CONFIG_ENV_SIZE, up_data)) {
 		        goto perror;
 		}
-#endif /* CONFIG_ARM9_CPU */
-
     	}
 #endif /* (CONFIG_ENV_SECT_SIZE_FLASH > CONFIG_ENV_SIZE_FLASH) */
 	puts("done\n");
@@ -420,11 +311,7 @@ done:
 #endif /* CMD_SAVEENV */
 
 #endif /* CONFIG_ENV_ADDR_REDUND */
-#ifdef CONFIG_ARM9_CPU
-void env_relocate_spec_flash(void)
-#else
 void env_relocate_spec(void)
-#endif /* CONFIG_ARM9_CPU */
 {
 #ifdef CONFIG_ENV_ADDR_REDUND
 	if (gd->env_addr != (ulong)&(flash_addr->data)) {
@@ -439,11 +326,7 @@ void env_relocate_spec(void)
 	}
 
 	if (flash_addr_new->flags != OBSOLETE_FLAG &&
-#ifdef CONFIG_ARM9_CPU
-            crc32(0, flash_addr_new->data, ENV_SIZE_FLASH) == flash_addr_new->crc) {
-#else
 	    crc32(0, flash_addr_new->data, ENV_SIZE) == flash_addr_new->crc) {
-#endif /* CONFIG_ARM9_CPU */
 		char flag = OBSOLETE_FLAG;
 
 		gd->env_valid = 2;
diff --git a/common/env_skov_compat.c b/common/env_skov_compat.c
deleted file mode 100644
index 8d988c3..0000000
--- a/common/env_skov_compat.c
+++ /dev/null
@@ -1,89 +0,0 @@
-/**
- * @file
- * @author Martin bjærre <mab@rosetechnology.dk>
- */
-#include <common.h>
-#include <environment.h>
-
-
-static char has_dataflash = 0;
-extern char * env_name_spec_dataflash;
-extern char * env_name_spec_flash;
-
-
-char *env_name_spec;
-env_t *env_ptr = (env_t *)CONFIG_ENV_ADDR_FLASH;
-
-
-void set_has_dataflash(int found)
-{
-	if (found == 1)
-	{
-		has_dataflash = 1;
-		env_name_spec = env_name_spec_dataflash;
-	}
-	else
-	{
-		has_dataflash = 0;
-		env_name_spec = env_name_spec_flash;
-	}
-}
-
-int do_we_have_dataflash(void)
-{
-	return has_dataflash;
-}
-
-
-//extern uchar env_get_char_spec_flash (int index);
-extern uchar env_get_char_spec_dataflash (int index);
-uchar env_get_char_spec (int index)
-{
-	if (has_dataflash)
-	{
-		return env_get_char_spec_dataflash(index);
-	}
-}
-
-extern void env_relocate_spec_flash(void);
-extern void env_relocate_spec_dataflash(void);
-void env_relocate_spec (void)
-{
-	if (has_dataflash)
-	{
-		env_relocate_spec_dataflash();
-	}
-	else
-	{
-		env_relocate_spec_flash();
-	}
-}
-
-extern int saveenv_flash(void);
-extern int saveenv_dataflash(void);
-int saveenv(void)
-{
-	if (has_dataflash == 1)
-	{
-		return saveenv_dataflash();
-	}
-	else
-	{
-		return saveenv_flash();
-	}
-}
-
-extern int env_init_flash(void);
-extern int env_init_dataflash(void);
-int env_init(void)
-{
-	if (has_dataflash)
-	{
-		return env_init_dataflash();
-	}
-	else
-	{
-		return env_init_flash();
-	}
-}
-
diff --git a/common/main.c b/common/main.c
index dd5028d..b8293b1 100644
--- a/common/main.c
+++ b/common/main.c
@@ -40,14 +40,6 @@
 #include <hush.h>
 #endif
 
-/**
- * @author Martin Bjærre <mab@rosetechnology.dk>
- * Rose add:
- */
-#ifdef CONFIG_ARM9_CPU
-extern int do_we_have_dataflash(void);
-#endif
-
 #include <post.h>
 #include <linux/ctype.h>
 #include <menu.h>
@@ -227,20 +219,6 @@ static inline
 int abortboot(int bootdelay)
 {
 	int abort = 0;
-/**
- * @author Martin Bjærre <mab@rosetechnology.dk>
- * Set env variable if we have nand
- */
-#ifdef CONFIG_ARM9_CPU
-	if (do_we_have_dataflash() == 1)
-	{
-		setenv("hasnand", "true");
-	}
-	else
-	{
-		setenv("hasnand", "false");
-	}
-#endif
 
 #ifdef CONFIG_MENUPROMPT
 	printf(CONFIG_MENUPROMPT);
diff --git a/drivers/mtd/cfi_flash.c b/drivers/mtd/cfi_flash.c
index b8edf37..43140f3 100644
--- a/drivers/mtd/cfi_flash.c
+++ b/drivers/mtd/cfi_flash.c
@@ -41,7 +41,7 @@
 #include <environment.h>
 #include <mtd/cfi_flash.h>
 #include <watchdog.h>
-#include <flash.h> /**< Søren Andersen <san@rosetechnology.dk> */
+
 /*
  * This file implements a Common Flash Interface (CFI) driver for
  * U-Boot.
@@ -2247,16 +2247,11 @@ void flash_protect_default(void)
 	/* Environment protection ON by default */
 #ifdef CONFIG_ENV_IS_IN_FLASH
 	flash_protect(FLAG_PROTECT_SET,
-#if CONFIG_ARM9_CPU
-                       CONFIG_ENV_ADDR_FLASH,
-                       CONFIG_ENV_ADDR_FLASH + CONFIG_ENV_SECT_SIZE_FLASH - 1,
-                       flash_get_info(CONFIG_ENV_ADDR_FLASH));
-#else
 		       CONFIG_ENV_ADDR,
 		       CONFIG_ENV_ADDR + CONFIG_ENV_SECT_SIZE - 1,
 		       flash_get_info(CONFIG_ENV_ADDR));
 #endif
-#endif
+
 	/* Redundant environment protection ON by default */
 #ifdef CONFIG_ENV_ADDR_REDUND
 	flash_protect(FLAG_PROTECT_SET,
diff --git a/drivers/mtd/dataflash.c b/drivers/mtd/dataflash.c
index 7f4446e..981ccd5 100644
--- a/drivers/mtd/dataflash.c
+++ b/drivers/mtd/dataflash.c
@@ -42,7 +42,6 @@ int AT91F_DataflashInit (void)
 	int found[CONFIG_SYS_MAX_DATAFLASH_BANKS];
 	unsigned char protected;
 
-//	printf("Looking for Dataflash\n"); /**< Martin Bjærre <mab@rosetechnology.dk> */
 	AT91F_SpiInit ();
 
 	for (i = 0; i < CONFIG_SYS_MAX_DATAFLASH_BANKS; i++) {
@@ -66,21 +65,6 @@ int AT91F_DataflashInit (void)
 			found[i] += dfcode;;
 			break;
 
-		case AT45DB041:
-			dataflash_info[i].Device.pages_number = 2048;
-			dataflash_info[i].Device.pages_size = 264;
-			dataflash_info[i].Device.page_offset = 9;
-			dataflash_info[i].Device.byte_mask = 0x300;
-			dataflash_info[i].Device.cs = cs[i].cs;
-			dataflash_info[i].Desc.DataFlash_state = IDLE;
-			dataflash_info[i].logical_address = cs[i].addr;
-			dataflash_info[i].id = dfcode;
-			found[i] += dfcode;;
-			#if defined(CONFIG_ARM9_CPU)
-			set_has_dataflash(1); /**< Martin Bjærre: used to determine if env is in nor or dataflash */
-			#endif
-			break;	
-
 		case AT45DB081:
 			dataflash_info[i].Device.pages_number = 4096;
 			dataflash_info[i].Device.pages_size = 264;
@@ -142,11 +126,7 @@ int AT91F_DataflashInit (void)
 			break;
 
 		default:
-			printf("Dataflash not recon... (dfcode=%xh)\n",dfcode);
 			dfcode = 0;
-			#if defined(CONFIG_ARM9_CPU)
-			set_has_dataflash(0); /**< Martin Bjærre: used to determine if env is in nor or dataflash */
-			#endif
 			break;
 		}
 		/* set the last area end to the dataflash size*/
@@ -221,14 +201,6 @@ void dataflash_print_info (void)
 			case AT45DB021:
 				printf("AT45DB021\n");
 				break;
-
-			case AT45DB041:
-				printf("AT45DB041\n");
-				break;
-
-			case AT45DB081:
-				printf("AT45DB081\n");
-
 			case AT45DB161:
 				printf("AT45DB161\n");
 				break;
@@ -240,7 +212,6 @@ void dataflash_print_info (void)
 			case AT45DB642:
 				printf("AT45DB642\n");
 				break;
-
 			case AT45DB128:
 				printf("AT45DB128\n");
 				break;
diff --git a/include/configs/arm9-cpu.h b/include/configs/arm9-cpu.h
index 1809414..e4bee5b 100644
--- a/include/configs/arm9-cpu.h
+++ b/include/configs/arm9-cpu.h
@@ -5,6 +5,8 @@
 #ifndef __CONFIG_H
 #define __CONFIG_H
 
+#define DEBUG 1
+
 #define CONFIG_AT91_LEGACY
 /*
  * SoC must be defined first, before hardware.h is included.
@@ -113,7 +115,7 @@
 
 #define CONFIG_CMD_PING		1
 #define CONFIG_CMD_DHCP		1
-#define CONFIG_CMD_NAND		1
+//#define CONFIG_CMD_NAND		1
 #define CONFIG_CMD_USB		1
 
 /** 
@@ -149,15 +151,15 @@
 	(ATMEL_BASE_SRAM1 + 0x1000 - GENERATED_GBL_DATA_SIZE)
 
 /* DataFlash */
-#define CONFIG_ATMEL_DATAFLASH_SPI
-#define CONFIG_HAS_DATAFLASH		1
-#define CONFIG_SYS_SPI_WRITE_TOUT		(5*CONFIG_SYS_HZ)
-#define CONFIG_SYS_MAX_DATAFLASH_BANKS		1
-#define CONFIG_SYS_DATAFLASH_LOGIC_ADDR_CS0	0xC0000000	/* CS0 */
+// #define CONFIG_ATMEL_DATAFLASH_SPI
+// #define CONFIG_HAS_DATAFLASH		1
+// #define CONFIG_SYS_SPI_WRITE_TOUT		(5*CONFIG_SYS_HZ)
+// #define CONFIG_SYS_MAX_DATAFLASH_BANKS		1
+// #define CONFIG_SYS_DATAFLASH_LOGIC_ADDR_CS0	0xC0000000	/* CS0 */
 
-#define AT91_SPI_CLK			15000000
-#define DATAFLASH_TCSS			(0x1a << 16)
-#define DATAFLASH_TCHS			(0x1 << 24)
+// #define AT91_SPI_CLK			15000000
+// #define DATAFLASH_TCSS			(0x1a << 16)
+// #define DATAFLASH_TCHS			(0x1 << 24)
 
 /* NOR flash, if populated */
 #define CONFIG_SYS_FLASH_CFI			1
@@ -172,15 +174,15 @@
 #define CONFIG_SYS_MONITOR_BASE	CONFIG_SYS_FLASH_BASE
 #define CONFIG_SYS_MONITOR_LEN	(256 << 10)
 
-#define CONFIG_ENV_IS_IN_DATAFLASH
+// #define CONFIG_ENV_IS_IN_DATAFLASH
 
 #define CONFIG_ENV_IS_IN_FLASH
-#define CONFIG_ENV_SIZE                         (16 * 1024)
-#define CONFIG_ENV_ADDR                         (CONFIG_SYS_DATAFLASH_LOGIC_ADDR_CS0 + 0x0007FE00)
+// #define CONFIG_ENV_SIZE                         (16 * 1024)
+// #define CONFIG_ENV_ADDR                         (CONFIG_SYS_DATAFLASH_LOGIC_ADDR_CS0 + 0x0007FE00)
 /* Address and size of Primary Environment Sector */
-#define CONFIG_ENV_SIZE_FLASH			0x000004000 /**< Must be smaller that 64KB otherwise h_export fails*/
-#define CONFIG_ENV_ADDR_FLASH                   (CONFIG_SYS_FLASH_BASE + 0x00080000)
-#define CONFIG_ENV_SECT_SIZE_FLASH		0x00020000 /**< One sector is 128KB in Nor flash*/
+#define CONFIG_ENV_SIZE			0x000004000 /**< Must be smaller that 64KB otherwise h_export fails*/
+#define CONFIG_ENV_ADDR                   (CONFIG_SYS_FLASH_BASE + 0x00080000)
+#define CONFIG_ENV_SECT_SIZE		0x00020000 /**< One sector is 128KB in Nor flash*/
 #define CONFIG_ENV_OVERWRITE			1 /**< San Rose It Shall be possible to change eth addr */
 
 #define MTDIDS_DEFAULT                  "nor0=physmap-flash.0"
@@ -228,8 +230,7 @@ KJP: layout_nor added 13 spaces to compensate for binary replacement by old layo
     "lcd_size=160 160\0" \
     "lcdtype=matrix\0" \
     "lcd_bglight=80\0"
-/* NAND flash */
-#ifdef CONFIG_CMD_NAND
+
 #define CONFIG_CMD_UBI
 #define CONFIG_CMD_UBIFS
 #define CONFIG_CMD_MTDPARTS
@@ -238,6 +239,9 @@ KJP: layout_nor added 13 spaces to compensate for binary replacement by old layo
 #define CONFIG_MTD_DEVICE
 #define CONFIG_MTD_PARTITIONS
 #define CONFIG_NAND_ATMEL
+
+/* NAND flash */
+#ifdef CONFIG_CMD_NAND
 #define CONFIG_SYS_MAX_NAND_DEVICE		1
 #define CONFIG_SYS_NAND_BASE			ATMEL_BASE_CS3
 #define CONFIG_SYS_NAND_DBW_8			1
@@ -285,7 +289,7 @@ KJP: layout_nor added 13 spaces to compensate for binary replacement by old layo
 #define CONFIG_SYS_MEMTEST_END			0x23800000
 
 
-#define CONFIG_SYS_PROMPT		"SKOV A/S> "
+#define CONFIG_SYS_PROMPT		"boot> "
 #define CONFIG_SYS_CBSIZE		256
 #define CONFIG_SYS_MAXARGS		16
 #define CONFIG_SYS_PBSIZE		(CONFIG_SYS_CBSIZE + sizeof(CONFIG_SYS_PROMPT) + 16)
diff --git a/include/environment.h b/include/environment.h
index da20161..6a21eb3 100644
--- a/include/environment.h
+++ b/include/environment.h
@@ -38,48 +38,27 @@
  */
 
 #if defined(CONFIG_ENV_IS_IN_FLASH)
-#if CONFIG_ARM9_CPU
-# ifndef	CONFIG_ENV_ADDR_FLASH
-#  define	CONFIG_ENV_ADDR_FLASH	(CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET_FLASH)
-# endif
-# ifndef	CONFIG_ENV_OFFSET_FLASH
-#  define	CONFIG_ENV_OFFSET_FLASH (CONFIG_ENV_ADDR_FLASH - CONFIG_SYS_FLASH_BASE_FLASH)
-# endif
-#else /* CONFIG_ARM9_CPU */
 # ifndef        CONFIG_ENV_ADDR
 #  define       CONFIG_ENV_ADDR (CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET)
 # endif
 # ifndef        CONFIG_ENV_OFFSET
 #  define       CONFIG_ENV_OFFSET (CONFIG_ENV_ADDR - CONFIG_SYS_FLASH_BASE)
 # endif
-#endif /* CONFIG_ARM9_CPU */
 # if !defined(CONFIG_ENV_ADDR_REDUND) && defined(CONFIG_ENV_OFFSET_REDUND)
 #  define	CONFIG_ENV_ADDR_REDUND	\
 		(CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET_REDUND)
 # endif
 # if defined(CONFIG_ENV_SECT_SIZE) || defined(CONFIG_ENV_SIZE)
 #  ifndef	CONFIG_ENV_SECT_SIZE
-#    define	CONFIG_ENV_SECT_SIZE	CONFIG_ENV_SIZE
+#   define CONFIG_ENV_SECT_SIZE CONFIG_ENV_SIZE
 #  endif
 #  ifndef	CONFIG_ENV_SIZE
 #   define	CONFIG_ENV_SIZE	CONFIG_ENV_SECT_SIZE
 #  endif
 # else
-#  error "Both CONFIG_ENV_SECT_SIZE and CONFIG_ENV_SIZE_FLASH undefined"
+#  error "Both CONFIG_ENV_SECT_SIZE and CONFIG_ENV_SIZE undefined"
 # endif
 
-# if defined(CONFIG_ENV_SECT_SIZE_FLASH) || defined(CONFIG_ENV_SIZE_FLASH)
-#  ifndef       CONFIG_ENV_SECT_SIZE_FLASH
-#    define     CONFIG_ENV_SECT_SIZE_FLASH    CONFIG_ENV_SIZE_FLASH
-#  endif
-#  ifndef       CONFIG_ENV_SIZE_FLASH
-#   define      CONFIG_ENV_SIZE_FLASH CONFIG_ENV_SECT_SIZE_FLASH
-#  endif
-# else  
-#  error "Both CONFIG_ENV_SECT_SIZE and CONFIG_ENV_SIZE_FLASH undefined"
-# endif
-
-
 # if defined(CONFIG_ENV_ADDR_REDUND) && !defined(CONFIG_ENV_SIZE_REDUND)
 #  define CONFIG_ENV_SIZE_REDUND	CONFIG_ENV_SIZE
 # endif
diff --git a/tools/envcrc.c b/tools/envcrc.c
index cf33149..08d2a8f 100644
--- a/tools/envcrc.c
+++ b/tools/envcrc.c
@@ -36,21 +36,12 @@
 #undef	__ASSEMBLY__
 
 #if defined(CONFIG_ENV_IS_IN_FLASH)
-#ifdef CONFIG_ARM9_CPU
-# ifndef  CONFIG_ENV_ADDR_FLASH
-#  define CONFIG_ENV_ADDR_FLASH       (CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET)
-# endif
-# ifndef  CONFIG_ENV_OFFSET_FLASH
-#  define CONFIG_ENV_OFFSET_FLASH (CONFIG_ENV_ADDR - CONFIG_SYS_FLASH_BASE)
-# endif
-#else
 # ifndef  CONFIG_ENV_ADDR
 #  define CONFIG_ENV_ADDR	(CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET)
 # endif
 # ifndef  CONFIG_ENV_OFFSET
 #  define CONFIG_ENV_OFFSET (CONFIG_ENV_ADDR - CONFIG_SYS_FLASH_BASE)
 # endif
-#endif /* CONFIG_ARM9_CPU */
 # if !defined(CONFIG_ENV_ADDR_REDUND) && defined(CONFIG_ENV_OFFSET_REDUND)
 #  define CONFIG_ENV_ADDR_REDUND	(CONFIG_SYS_FLASH_BASE + CONFIG_ENV_OFFSET_REDUND)
 # endif
-- 
1.8.3.1

